source/opt/debug_info_manager.h

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SOURCE_OPT_DEBUG_INFO_MANAGER_H_
+#define SOURCE_OPT_DEBUG_INFO_MANAGER_H_
+
+#include <unordered_map>
+
+#include "source/opt/instruction.h"
+#include "source/opt/module.h"
+
+namespace spvtools {
+namespace opt {
+namespace analysis {
+
+// A class for analyzing, managing, and creating OpenCL.DebugInfo.100 extension
+// instructions.
+class DebugInfoManager {
+ public:
+  DebugInfoManager(IRContext* context);
+  // Constructs a def-use manager from the given |module|. All internal messages
+  // will be communicated to the outside via the given message |consumer|. This
+  // instance only keeps a reference to the |consumer|, so the |consumer| should
+  // outlive this instance.
+
+  DebugInfoManager(const DebugInfoManager&) = delete;
+  DebugInfoManager(DebugInfoManager&&) = delete;
+  DebugInfoManager& operator=(const DebugInfoManager&) = delete;
+  DebugInfoManager& operator=(DebugInfoManager&&) = delete;
+
+  friend bool operator==(const DebugInfoManager&, const DebugInfoManager&);
+  friend bool operator!=(const DebugInfoManager& lhs,
+                         const DebugInfoManager& rhs) {
+    return !(lhs == rhs);
+  }
+
+  // Clones DebugDeclare or DebugValue for a local variable whose result id is
+  // |orig_var_id|. Set Variable operand of the new DebugDeclare as
+  // |new_var_id|. Return the new DebugDeclare or DebugValue. If this is the
+  // first DebugDeclare or DebugValue, keep it in |local_var_id_to_dbgdecl_|.
+  Instruction* CloneDebugDeclare(uint32_t orig_var_id, uint32_t new_var_id);
+
+  // Returns id of new DebugInlinedAt. Its Line operand is the line number
+  // of |line| if |line| is not nullptr. Otherwise, its Line operand
+  // is the line number of lexical scope of |scope|. Its Scope and Inlined
+  // operands are Scope and Inlined of |scope|. Note that this function puts
+  // the new DebugInlinedAt into the tail of debug instructions.
+  uint32_t CreateDebugInlinedAt(const Instruction* line,
+                                const DebugScope& scope);
+
+  // Creates a new DebugInfoNone instruction and returns it. In addition,
+  // insert the new DebugInfoNone instruction before the head of debug
+  // instructions.
```

Comment:
I still this should return an already existing instruction.  

The promise of the location should be removed unless you have a specific reason.  I could see a requirement that it be at the start to be useful to ensure that it can be referenced in any other debug instruction.

If you keep it, please reword to "The returned instruction will be at the start of the debug instructions."